build_index.py

```diff
@@ -0,0 +1,170 @@
+    registry = toml.load("registry.toml")
+    # Get authentication for Github API
+    if os.getenv('CI'):
+        user = os.getenv('GITHUB_ACTOR')
+        tkn = os.getenv('GITHUB_TOKEN')
+    else:
+        account = toml.load("account.toml")
+        user = account["github"]["user"]
+        tkn = account["github"]["token"]
+
+    # Setup and authenticate API objects
+    api_context = {"github": github3.login(user, tkn), 
+                "gitlab": GitlabLite()}
+
+    # Loop over packages from registry.toml
+    n_registered = 0
+    n_update = 0
+    n_new = 0
+    n_skip = 0
+    n_breaking = 0
+    n_failed = 0
+    for pkg_name in registry:
+
+        for pkg_version in registry[pkg_name]:
+
+            n_registered += 1
+
+            try:
+
+                # Check registry toml entry
+                pkg_info = registry[pkg_name][pkg_version]
+                check_registry_entry(pkg_name,pkg_version,pkg_info)
+
+                # We can skip versioned packages if already in index.json
+                #  and git tags match
+                if (f"{pkg_name}" in index["packages"]
+                    and f"{pkg_version}" in index["packages"][pkg_name]
+                    and "git-tag" in index["packages"][pkg_name][pkg_version]
+                    and "tag" in registry[pkg_name][pkg_version]
+                    and index["packages"][pkg_name][pkg_version]["git-tag"] ==
+                        registry[pkg_name][pkg_version]["tag"]):
+
+                    print("        Package version already indexed and up-to-date, skipping.")
+                    n_skip += 1
+                    continue
+                
+                # Fetch fpm.toml for package version
+                if "tag" in pkg_info:
+                    fpm_toml = fetch_fpm_toml(api_context,
+                                              pkg_info["git"],pkg_info["tag"])
+                    ref = pkg_info['tag']
+                else:
+                    fpm_toml = fetch_fpm_toml(api_context,pkg_info["git"])
+                    ref = None
+
+                # Build json index entry
+                pkg_info_full = check_fpm_toml(fpm_toml)
+                pkg_info_full["git"] = pkg_info["git"]
+                pkg_info_full["git-tag"] = ref
+
+                # Counting
+                if (f"{pkg_name}" in index["packages"]
+                    and f"{pkg_version}" in index["packages"][pkg_name]):
+
+                    n_update += 1
+                else:
+                    n_new += 1
+                
+                # Save to index dict
+                if f"{pkg_name}" not in index["packages"]:
+                    index["packages"][pkg_name] = {}
+
+                index["packages"][pkg_name][pkg_version] = pkg_info_full
+
+            except Exception:
+                print(f"        (!) Error processing package '"
+                     +f"{pkg_name}-{pkg_version}, skipping.\n")
+                traceback.print_exc()
+                print("")
```

Comment:
This is fine for now, later it can be improved to catch specific exceptions. In general one should not catch "all exceptions", as it can hide bugs in our code that we want to know about.